File: src/Trefunge98Strategy.java
/**
 * @file Trefunge98Strategy.java
 * Strategy for Trefunge-98 commands.
 * @author Conlan Wesson
 */


class Trefunge98Strategy extends Befunge98Strategy{
	public Trefunge98Strategy(Field f, InstructionPointer i, StackStack s){
		super(f, i, s);
	}
	
	public boolean execute(char cmd){
		switch(cmd){
			case 'h':
				ip.setDir(0, 0, -1);
				break;
			case 'l':
				ip.setDir(0, 0, 1);
				break;
			case 'm':
				if(stack.top().pop() != 0){
					ip.setDir(0, 0, -1);
				}else{
					ip.setDir(0, 0, 1);
				}
				break;
			case '?':{
				int r = rnd.nextInt(6);
				switch(r){
					case 0:
						ip.setDir(0, 1, 0); break;
					case 1:
						ip.setDir(-1, 0, 0); break;
					case 2:
						ip.setDir(1, 0, 0); break;
					case 3:
						ip.setDir(0, -1, 0); break;
					case 4:
						ip.setDir(0, 0, 1); break;
					case 5:
						ip.setDir(0, 0, -1); break;
				}
				break;
			}
			
			default:
				return super.execute(cmd);
		}
		return true;
	}
}

File: src/Trefunge98Strategy.cpp
/**
 * @file Trefunge98Strategy.cpp
 * Strategy for Trefunge-98 commands.
 * @author Conlan Wesson
 */

#include "Trefunge98Strategy.h"
#include "FungeRunner.h"
#include "FungeUniverse.h"

namespace Funge {


Trefunge98Strategy::Trefunge98Strategy(FungeRunner& r) :
	FungeStrategy(r)
{
	r.pushSemantic('h', std::bind(&Trefunge98Strategy::instructionHigh, this), FungeSemantic::MOVEMENT);
	r.pushSemantic('l', std::bind(&Trefunge98Strategy::instructionLow, this), FungeSemantic::MOVEMENT);
	r.pushSemantic('m', std::bind(&Trefunge98Strategy::instructionIf, this));
}

FungeError Trefunge98Strategy::instructionHigh(){
	if(!runner.getUniverse().invertHL()){
		ip.setDelta(Vector{0, 0, 1});
	}else{
		ip.setDelta(Vector{0, 0, -1});
	}
	return ERROR_NONE;
}

FungeError Trefunge98Strategy::instructionLow(){
	if(!runner.getUniverse().invertHL()){
		ip.setDelta(Vector{0, 0, -1});
	}else{
		ip.setDelta(Vector{0, 0, 1});
	}
	return ERROR_NONE;
}

FungeError Trefunge98Strategy::instructionIf(){
	if(stack.top().pop() == 0){
		ip.setDelta(Vector{0, 0, 1});
	}else{
		ip.setDelta(Vector{0, 0, -1});
	}
	return ERROR_NONE;
}

FungeStrategy* Trefunge98Strategy::clone(FungeRunner& r) const{
	return new Trefunge98Strategy(r);
}

}
